### PR TITLE
Implemented Dummy Class

### LINKED ISSUE #1283: Add support for Dummy.
SymPy has Dummy symbols and is used in many places.

### PR DESCRIPTION
**Pull request by ShikharJ**
Fixes #1283

### CODE REVIEW

File: symengine/tests/basic/test_functions.cpp [hunk 1]
```diff
@@ -4062,3 +4063,17 @@ TEST_CASE("min: functions", "[functions]")
...
+    RCP<const Symbol> x1 = symbol("x");
+    RCP<const Symbol> x2 = symbol("x");
+    RCP<const Symbol> xdummy1 = dummy("x");
+    RCP<const Symbol> xdummy2 = dummy("x");
+
+    CHECK(eq(*x1, *x2));
+    CHECK(neq(*x1, *xdummy1));
+    CHECK(neq(*xdummy1, *xdummy2));
+    CHECK(neq(*dummy(), *dummy()));
+    CHECK(neq(*dummy("x"), *dummy("x")));
```
Comment (ShikharJ) [inline symengine/tests/basic/test_functions.cpp]: @isuruf I'm unable to understand why the above line passes while `4076` and `4077` fail.

File: symengine/symbol.h [hunk 2]
```diff
@@ -46,12 +46,53 @@ class Symbol : public Basic
     }
 };
 
+class Dummy : public Symbol
+{
+private:
+    //! Dummy count
+    std::size_t count_;
```
Comment (isuruf) [inline symengine/symbol.h]: This should be a static variable for the class. See SymPy implementation.

File: symengine/symbol.h [hunk 3]
```diff
@@ -46,12 +46,53 @@ class Symbol : public Basic
     }
 };
 
+class Dummy : public Symbol
+{
+private:
+    //! Dummy count
+    static size_t count_;
```
Comment (isuruf) [inline symengine/symbol.h]: `count_` is common to all the objects of class `Dummy`. You'll also need a `dummy_index` per object.

**Comment by ShikharJ** [comment]
@isuruf Why are these tests failing?
```
-------------------------------------------------------------------------------
And, Or : Basic
-------------------------------------------------------------------------------
/home/shikhar/symengine/symengine/tests/logic/test_logic.cpp:122
...............................................................................

/home/shikhar/symengine/symengine/tests/logic/test_logic.cpp:152: FAILED:
  REQUIRE( unified_eq(v, u) )
with expansion:
  false

-------------------------------------------------------------------------------
Xor : Basic
-------------------------------------------------------------------------------
/home/shikhar/symengine/symengine/tests/logic/test_logic.cpp:251
...............................................................................

/home/shikhar/symengine/symengine/tests/logic/test_logic.cpp:272: FAILED:
  REQUIRE( unified_eq(v, u) )
with expansion:
  false

===============================================================================
test cases:  9 |  7 passed | 2 failed
assertions: 71 | 69 passed | 2 failed
```

**Comment by isuruf** [comment]
Those tests are wrong. It should have used `vec_basic_eq_perm` instead of `unified_eq`.
When you added the Dummy symbol, the hashes changed and therefore the order changed.

**Comment by ShikharJ** [comment]
Ping @isuruf

File: symengine/symbol.cpp [hunk 4]
```diff
@@ -32,4 +41,50 @@ int Symbol::compare(const Basic &o) const
...
+size_t Dummy::count_ = 0;
+
+Dummy::Dummy() : Symbol("Dummy_" + to_string(count_))
+{
+    SYMENGINE_ASSIGN_TYPEID()
+    count_ += 1;
+    dummy_index = count_;
+}
+
+Dummy::Dummy(const std::string &name) : Symbol(name)
```
Comment (isuruf) [inline symengine/symbol.cpp]: In sympy, a "_" is prepended the name. `Dummy("x")` outputs `_x`.

**Pull request merged**